subt.test_subt: add test_go_to_entrance

zwn opened the pull request:
I'd like to eventually improve trace following robustness (taking shortcuts vs getting stuck) and also wall following (to weed out the situations where the robot spins in place, that we have experienced on both real and simulated robots). I'd also like to continue my work on monitors (timeouts and such). For all of that I feel a strong need to have some tests where I can deterministically reproduce the problem.

This is the first step towards that goals.

m3d commented on subt/test_subt.py:
```diff
@@ -1,7 +1,173 @@
+        return dt
+
+
+class Simulation:
```
this part seems too big  ... I would probably keep it separate from unittest??

zwn replied:
yeah, that's right - any idea about the name? `subt/simulation.py`?

m3d replied:
that sounds reasonable ... i.e. yes

zwn commented:
I will move the PYTHONIOENCODING stuff to a separate PR (the tests only failed on CI, it runs on both Linux and Windows ok).

zwn commented:
I have moved the stuff with encoding to #479 and changed base branch of this PR to be based on that.

jisa commented on subt/simulation.py:
```diff
@@ -0,0 +1,159 @@
+    print(stylize(f"{caller_class.__name__}.{caller}", fg(color_name)), *args)
+
+
+def clip(min_value, value, max_value):
```
Not sure how you feel about external dependencies at this point. This is np.clip() and you may not need to define it yourself.

jisa commented on subt/simulation.py:
```diff
@@ -0,0 +1,159 @@
+        diff_forward = clip(-0.1, forward - self.speed_forward, 0.1)
+        diff_angular = clip(-10.0, angular - self.speed_angular, 10.0)
+        #print(diff_forward, diff_angular)
+        self.speed_forward += diff_forward
```
diff_forward seems to basically be acceleration, but the unit is m/s/function_call? What do you think about somehow turning it into m/s/s? Similar for the rotation.

jisa commented on subt/simulation.py:
```diff
@@ -0,0 +1,159 @@
+    @color.sea_green_3
+    def on_desired_speed(self, dt, channel, data):
+        forward = data[0]/1000.
+        angular = data[1]/100.
```
How do you work with angles in the rest of Osgar? In my experience, mixing degrees and radians is very dangerous. This one seems to be degrees, but it is not obvious. Do you always convert to one representation, or is it a mixture depending on what is being done at the given point?

zwn replied:
It is a mixture everywhere. I've added some comments plus converted all calculation to SI units to make it clearer.

jisa commented on subt/simulation.py:
```diff
@@ -0,0 +1,159 @@
+from osgar.lib import quaternion
+
+try:
+    from colored import stylize, fg
```
How important is this coloring? There seems to be lots of boilerplate and non-obvious code related to it.

zwn replied:
It made the output much more readable but I agree that what it added to readability of the output, it removed from the readability of the actual code. Some day I'd like to have something like it, but probably not today.

jisa commented on subt/simulation.py:
```diff
@@ -0,0 +1,159 @@
+_saved_print = print
+
+def verbose(ok):
+    print(ok)
```
Is this a left-over debugging print?

jisa commented on subt/simulation.py:
```diff
@@ -0,0 +1,159 @@
+    @color.gold_1
+    def on_request_origin(self, dt, channel, data):
+        log(dt, channel, data)
+        # start sending origin data until robot arrives within 0.3m from (0,0,0)
```
Where are the 0.3 and (0, 0, 2) parts in the code?

zwn replied:
That is actually forgotten TODO. It is how ros_proxy_node.cc behaves. It is here

https://github.com/robotika/osgar/blob/be3010c2b0a268721b990cdb07cc5dc823e73dfa/subt/ros/proxy/ros_proxy_node.cc#L416-L421

But the current code in the simulation sends the origin only once (actually twice - first to get robot name, second to get the origin). The SubT application does not require the origin to be sent more than once. Sending it many times probably should help remove any positional error accumulated in the start area. I have reworded the comment to acknowledge the difference for now.

zwn commented:
@jisa: Thanks for the thorough review. It helped me to make the code better. I've addressed all the comments. If I missed anything feel free to reopen the discussion or add a new one.
